neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,258 @@
+            weights=tf.transpose(self.train_mask)[:, :min_time])
+
+    @tensor
+    def runtime_logprobs(self) -> tf.Tensor:
+        return tf.nn.log_softmax(self.runtime_logits)
+
+    @property
+    def output_dimension(self) -> Union[int, tf.Tensor]:
+        raise NotImplementedError("Abstract property")
+
+    def get_initial_loop_state(self) -> LoopState:
+        raise NotImplementedError("Abstract method")
+
+    def loop_continue_criterion(self, *args) -> tf.Tensor:
+        loop_state = LoopState(*args)
+        finished = loop_state.feedables.finished
+        not_all_done = tf.logical_not(tf.reduce_all(finished))
+        before_max_len = tf.less(loop_state.feedables.step,
+                                 self.max_output_len)
+        return tf.logical_and(not_all_done, before_max_len)
+
+    def get_body(self, train_mode: bool, sample: bool = False) -> Callable:
+        raise NotImplementedError("Abstract method")
+
+    def finalize_loop(self, final_loop_state: LoopState,
+                      train_mode: bool) -> None:
+        pass
+
+    def decoding_loop(self, train_mode: bool, sample: bool = False) -> Tuple[
+            tf.Tensor, tf.Tensor, tf.Tensor]:
+
+        initial_loop_state = self.get_initial_loop_state()
+        final_loop_state = tf.while_loop(
+            self.loop_continue_criterion,
+            self.get_body(train_mode, sample),
+            initial_loop_state)
+
+        self.finalize_loop(final_loop_state, train_mode)
+
+        logits = final_loop_state.histories.logits.stack()
+        decoder_outputs = final_loop_state.histories.decoder_outputs.stack()
+
+        # TODO mask should include also the end symbol
+        mask = final_loop_state.histories.mask.stack()
+
+        return logits, decoder_outputs, mask
+
+    def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
+        """Populate the feed dictionary for the decoder object.
+
+        Arguments:
+            dataset: The dataset to use for the decoder.
+            train: Boolean flag, telling whether this is a training run
+        """
+        sentences = cast(Iterable[List[str]],
+                         dataset.get_series(self.data_id, allow_none=True))
+
+        if sentences is None and train:
+            raise ValueError("When training, you must feed "
+                             "reference sentences")
+
+        sentences_list = list(sentences) if sentences is not None else None
+
+        fd = {}  # type: FeedDict
+        fd[self.train_mode] = train
+
+        go_symbol_idx = self.vocabulary.get_word_index(START_TOKEN)
+        fd[self.go_symbols] = np.full([len(dataset)], go_symbol_idx,
+                                      dtype=np.int32)
+
+        if sentences is not None:
+            # train_mode=False, since we don't want to <unk>ize target words!
+            inputs, weights = self.vocabulary.sentences_to_tensor(
+                sentences_list, self.max_output_len, train_mode=False,
```

Comment:
udělat issue aby se změnil train_mode ve slovníku na něco jinýho, co víc odpovídá. třeba unk_replace nebo tak

Reply:
To nechám na tobě :)

Reply:
DONE